File: restdoc-rpc-client/restdoc-dubbo-client/src/main/kotlin/restdoc/client/dubbo/DubboInvokerImpl.java
package restdoc.client.dubbo;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.dubbo.common.utils.ReflectUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import restdoc.client.api.Invoker;
import restdoc.client.api.model.DubboInvocation;
import restdoc.client.api.model.DubboInvocationResult;
import restdoc.client.api.model.InvocationResult;
import restdoc.client.api.model.ObjectHolder;

import java.lang.reflect.Method;

/**
 * Dubbo invoker impl
 *
 * @see org.apache.dubbo.rpc.RpcInvocation
 */
@Component
public class DubboInvokerImpl implements Invoker<DubboInvocation> {
    private ObjectMapper mapper = new ObjectMapper();

    private final DubboRefBeanManager beanManager;

    @Autowired
    public DubboInvokerImpl(DubboRefBeanManager beanManager) {
        this.beanManager = beanManager;
    }

    @Override
    public InvocationResult rpcInvoke(DubboInvocation invocation) {
        Object bean = beanManager.getRefBean(invocation.getRefName());
        try {
            Class[] paramTypes = invocation.getParameters()
                    .stream()
                    .map(pa -> {
                        try {
                            return Class.forName(pa.className);
                        } catch (ClassNotFoundException e) {
                            e.printStackTrace();
                            throw new RuntimeException(e.getMessage());
                        }
                    }).toArray(Class[]::new);

            Object[] parameters = invocation.getParameters()
                    .stream()
                    .map(this::mappingValue)
                    .toArray(Object[]::new);

            Method method = bean.getClass().getMethod(invocation.getMethodName(), paramTypes);
            method.setAccessible(true);

            Object returnValue = method.invoke(bean, parameters);

            String serializedReturnValue = returnValue == null ?
                    "" : mapper.writeValueAsString(returnValue);
            return new DubboInvocationResult(
                    false,
                    null,
                    invocation,
                    serializedReturnValue,
                    method.getReturnType().toString());

        } catch (Exception e) {
            e.printStackTrace();
            return new DubboInvocationResult(
                    false,
                    e.getMessage(),
                    invocation,
                    "",
                    Void.class.toString());
        }
    }

    private Object mappingValue(ObjectHolder<Object> holder) {
        Class type = ReflectUtils.forName(holder.className);
        if (ReflectUtils.isPrimitive(type)) {
            return holder.value;
        } else {
            return mapper.convertValue(holder.value, type);
        }
    }

}

File: restdoc-rpc-client/restdoc-dubbo-client/src/main/kotlin/restdoc/client/dubbo/DubboInvokerImpl.kt
package restdoc.client.dubbo

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.dubbo.common.utils.ReflectUtils
import org.springframework.stereotype.Component
import restdoc.client.api.Invoker
import restdoc.client.api.model.DubboInvocation
import restdoc.client.api.model.DubboInvocationResult
import restdoc.client.api.model.ObjectHolder

/**
 * Dubbo invoker impl
 *
 * @sample org.apache.dubbo.rpc.RpcInvocation
 */
@Component
class DubboInvokerImpl(private val beanManager: DubboRefBeanManager) : Invoker<DubboInvocation> {

    private val mapper: ObjectMapper = ObjectMapper()

    override fun rpcInvoke(t: DubboInvocation): DubboInvocationResult {
        val bean = beanManager.getRefBean(t.refName)

        val paramTypes = t.parameters.map { Class.forName(it.className) }.toTypedArray()
        val params = t.parameters.map { mappingValue(it) }.toTypedArray()

        val method = bean.javaClass.getMethod(t.methodName, *paramTypes)
        method.isAccessible = true

        try {
            val returnValue: Any? = method.invoke(bean, *params)

            val serializedReturnValue = if (returnValue != null) mapper.writeValueAsString(returnValue)
            else ""

            return if (returnValue == null) {
                DubboInvocationResult(
                        isSuccessful = true,
                        exceptionMsg = null,
                        invocation = t,
                        returnValue = "",
                        returnValueType = method.returnType.toString())
            } else {
                DubboInvocationResult(
                        isSuccessful = false,
                        exceptionMsg = null,
                        invocation = t,
                        returnValue = serializedReturnValue,
                        returnValueType = method.returnType.toString())

            }
        } catch (e: Exception) {
            e.printStackTrace()
            return DubboInvocationResult(
                    isSuccessful = false,
                    exceptionMsg = e.message,
                    invocation = t,
                    returnValue = "",
                    returnValueType = method.returnType.toString())
        }
    }

    private fun mappingValue(holder: ObjectHolder<Any>): Any? {
        val type = ReflectUtils.forName(holder.className)
        if (ReflectUtils.isPrimitive(type)) {
            return holder.value
        } else {
            return mapper.convertValue(holder.value, type)
        }
    }
}